content/blog/django-model-fields-oracle-database-sequences/index.md

```diff
@@ -0,0 +1,221 @@
+
+
+# What about triggers?
+
+It's possible there is already a trigger on the table that will get/set the sequence when `NULL`:
+
+```sql
+-- example from: https://oracle-base.com/articles/misc/autonumber-and-identity
+CREATE OR REPLACE TRIGGER dept_bir
+BEFORE INSERT ON department
+FOR EACH ROW
+WHEN (new.department_id IS NULL)
+BEGIN
+  :new.department_id := dept_seq.NEXTVAL;
+END;
+/
+```
+
+That's awesome but it can also lead to some weird issues/errors in Django (e.g., admin showing "None" when saving). If the trigger is setup to check NULL no harm in setting it ourselves but you should always check triggers for any tables when setting up new models in Django.
+
+# Advanced model method decorator usage
+
+Let's say you have a table without a primary key and without `sequence`, `identity` or `auto_increment`:
+
+```sql
+CREATE TABLE dept_type (TYPE_KEY NUMBER(10), DESCR VARCHAR2(50));
+```
+
+You look at some of the code using it and see that the `TYPE_KEY` was being set using this SQL:
+
+```sql
+select 1+max(type_key) from dept_type;
+```
+
+Run `inspectdb` like we did above and add our new decorator to get the final model:
+
+```python
+# departments/models.py
+from django.db import models
+from myapp.models import set_sql_for_field
+
+class DepartmentType(models.Model):
+    type_key = models.IntegerField(primary_key=True)
+    descr = models.CharField(max_length=50)
+
+    class Meta:
+        db_table = 'dept_type'
+        managed = False
+
+    @set_sql_for_field('type_key', 'select 1+max(type_key) from dept_type')
+    def save(self, *args, **kwargs):
+        super().save(*args, **kwargs)
+```
+
+That's all there is to it. We set `primary_key` on the `type_key` field to tell Django not to use the automatic `id` field (see <https://docs.djangoproject.com/en/3.1/topics/db/models/#automatic-primary-key-fields>).
+
+# Additional model/field considerations
+
+Depending on your models you may want to set certain fields as read-only in the Django admin or Django REST Framework to avoid them being set directly:
+
+```python
+# departments/admin.py
+from django.contrib import admin
+from myapp.departments.models import DepartmentType
+
+@admin.register(DepartmentType)
+class DepartmentTypeAdmin(admin.ModelAdmin):
+    list_display = ('__str__', 'type_key', )
+    readonly_fields = ('type_key', )
+    search_fields = ('=type_key', 'descr', )
+
+
+# departments/serializers.py
+from rest_framework.serializers import ModelSerializer
+from myapp.departments.models import DepartmentType
+
+class DepartmentTypeSerializer(ModelSerializer):
+    class Meta:
+        fields = '__all__'
+        model = DepartmentType
+        read_only_fields = ['type_key']
+```
+
+# Closing thoughts
+
+**Whenever possible use primary keys and `AUTO_INCREMENT` or `IDENTITY`**. For the times you can't, hopefully something like `set_sql_for_field` can help!
```

Comment:
Why? If anything you're making the case that those aren't needed. There are valid reasons for using sequences, just googling sequences vs auto_increment will show lots of debate, and there is even the case for using uuid generation in postgres (which btw, your decorator will handle wonderfully.) We should probably backup a recommendation like this somehow.

Reply:
In Oracle `IDENTITY` is using sequences under the hood and we're not saying not to use sequences. I'm not sure I follow how we're making the case that they're not needed. The article is about something you can't change that maybe wasn't built in the best way and how to work around that.

The recommendation is to use `IDENTITY` as a convenience and to get some of the extra features/requirements for identity like only being able to use one of them on a table (more at https://docs.oracle.com/database/121/SQLRF/statements_7002.htm#SQLRF55661).

The intent of the wrap-up isn't to consider every possible way to assign an ID in every DBMS but to give a high-level "use the magic" suggestion.

Reply:
> The article is about something you can't change that maybe wasn't built in the best way and how to work around that.

Then maybe leave the recommendation off entirely?

Reply:
We're trying to educate and I don't agree that "use primary keys and IDs" is in any way controversial. I also think saying nothing isn't helpful at all.

We've spent a good bit of the article effectively saying, "the way it was built wasn't ideal" and it makes sense to close with a "here's how to do it right".

I've updated the PR, let me know if this feels a little better.

Reply:
Nothing here is controversial. The article is about dealing with legacy schema and that's great but that message is mixed with here's the way to do it right without anything backing up why we feel that's the right approach.

Reply:
If it's not controversial to suggest auto-incrementing primary keys as a sane default I'm not sure why we need to back it up. To me, that's been a longstanding best-practice.

Are you hoping to see links or something to why you should (generally) use them?

Reply:
We've talked a bit about this via Slack and we're cool with how it is. We might write another article with more detail about how to do things the right way.